README.md
# Alert-autoconf
## Alerting auto configuration

This Utility uses a configuration file (alert.yaml), and automatically creates triggers and alerts.

### Docker run: Apply create/change/delete commands for Triggers and Notifications

You need to be in the directory with the configuration file "./alerts.yaml"

```shell
docker run -v `pwd`:/conf registry.yourdomain.ru/alerting/alert:latest \
    --user $LOGIN --password $PASSWORD \
    --redis_token_storage "redis://$REDIS:6379/1"" \
    --token "hands::$SERVICE::$ENV" \
    --log-level DEBUG
    --cluster $CLUSTER
```
$LOGIN - Moira account. It is best to use special "system" account<br />
$PASSWORD - password<br />
$REDIS - redis host for internal metadata<br />
$SERVICE - the name of the service for which triggers and notifications will be created<br />
$ENV - prod/staging/dev<br />
$CLUSTER (optional) - We use this parameter when we have a new k8s cluster, <br />
                      and we want to deploy previously prepared triggers for it<br />

### Validation config file (alert.yaml):
```shell
docker run -v `pwd`:/conf registry.yourdomain.ru/alerting/alert-validator:latest \
    --log-level DEBUG
```
```shell
docker pull registry.yourdomain.ru/alerting/alert-validator:latest; \
find . -type f -iname 'alert.yaml' -print0 | \
xargs --null --max-args=1 --max-procs=1 --no-run-if-empty -I{} \
docker run -v `pwd`:/conf  registry.yourdomain.ru/alerting/alert-validator:latest\
    --log-level DEBUG \
    --config "{}"
```

### Tests run 

```shell
docker run -ti --entrypoint=make registry.yourdomain.ru/alerting/alert:latest test
```

### Config file describe
The configuration file consists of two required sections:

    "triggers:" - configuring triggers
    "alerting:" - setting of contacts for notification

```
---
version: "1"                  # version package "1" or "1.1"
prefix: "any-string-"         # a prefix is added for the names of triggers and tags
                              # (only for version: "1.1")

# Triggers section
triggers:
  - name: Trigger_1           # Trigger name
    targets:                  # Targets list
      - apps.services.my_service.api.query_rec_POST.request_time.200.count
      - apps.services.my_service.api.query_rec_POST.request_time.5*.count
    desc: Test description 1  
    dashboard: "http://grafana.yourdomain.ru/grafana/d/000000596/graphite-clickhouse?panelId=17&fullscreen"
                              # Link to a specific dashboard panel in Grafana
    is_pull_type: true        # Fetch data in Graphite without using Moira's internal storage. Flase by default.
    pending_interval:0        # The number of seconds that the trigger must spend in the new state 
                              # before Moira sends a notification on it
    warn_value: 15            # Execute mailing with the status' WARN
    error_value: 10           # Execute mailing with the status' ERROR 

                              # (If the value of the error_value field is greater than or equal to warn_value,
                              #  then the notification algorithm will be as follows: 
                              #  Execute 'Warning' or 'Error' - if the metric value is greater than the specified
                              # )

    ttl: 600                  # The number of seconds after which you need to make a mailing with the status from the 'ttl_state' field
    tags:                     # Subset of triggers
      - service-1
      - "cluster-{cluster}"   # Dynamic cluster name substitution (Optional!)
                              # (Strings that use {cluster} are recommended to be wrapped in quotes 
                              #  so that the yaml parser does not swear)
    ttl_state: OK             # OK | WARN | ERROR | NODATA | DEL
                              # All states are described here - http://moira.readthedocs.io/en/latest/user_guide/efficient.html

    parents:                  # List of parent triggers (optional)
      - name: "DC shutdown"   # The link to the parent is his name and a set of tags
        tags:
        - datacenters
        - global
      - name: "autoconf maintenance"
        tags:
        - autoconf

    expression: "t1 > 10 ? ERROR : (t2 > 4 ? WARN : OK)"
                              # Expression input.
                              # Supports govaluate and graphite syntax. (t1, t2, tN - in the order specified in targets)
                              # Described - http://moira.readthedocs.io/en/latest/user_guide/advanced.html

    time_start: "12:30"       # Start time of this trigger
    time_end: "23:59"         # End time of this trigger
    day_disable:              # Days in which triggers don't work (list)
      - Tue                   # Mon | Tue | Wed | Thu | Fri | Sat | Sun

  - name: Trigger_2
    targets:
      - movingAverage(apps.services.{cluster}.monitoring.modules.worker.*.handleMessage.time.count, 2)
    dashboard: grafana.yourdomain.ru/grafana/d/000000596/graphite-clickhouse?panelId=17&fullscreen
    is_pull_type: true
    ttl: 60
    ttl_state: ERROR
    expression: "t1 <= 0.3 ? ERROR : OK"
    tags:
      - sms
      - mail
      - slack


# Alerting section
alerting:
  # Simple notifications
  - tags:                            # List of tags used in trigger descriptions
      - service 1                    # It is enough to match one tag with a trigger tag
    time_start: "12:30"              # Application start time
    time_end: "23:50"                # Application end time
    day_disable:                     # Days in which this setting is ignored
      - Tue
    contacts:                        # List of available contacts
      - type: mail                   # type of notification. It is possible to specify the following values:
                                     # mail, send-sms, slack, jira
        value: qwe@qwe.qwe           # Contact value. The text in the field must be enclosed in quotation marks,
                                     # if special symbols #, @, +, etc. are used.

  # Notify with escalations
  - tags:
      - service_1
      - ERROR
    contacts:
      - type: slack
        value: #channel_service_1                 # Send a notification to the Slack channel of the service owners
      - type: send-sms
        value: +7123456789                        # Duty person service-1
    escalations:                                  # Escalation section. 
                                                  #  If it is described in the subscription, then all messages sent to Slack will have an "Acknoweledge" button
      - contacts:                                 # Escalation will be interrupted if the "Acknoweledge" button was pressed, 
          - type: jira                            #  or the trigger itself returned to the "OK" state during this time.
            value: monitoring_team_24x7
          - type: slack                           # Send a notification to the Slack team 24x7
            value: #channel_monitoring_team_24x7  
        offset_in_minutes: 5                      # Notification will be sent 5 minutes after the start

      - contacts:                     
          - type: send-sms             
            value: '+7123456789'                  # Team lead responsible for the service       
        offset_in_minutes: 10                     # Notifications will be sent 10 minutes after the start
```

---
Example configuration file:

 [valid_config.yaml](./tests/valid_config.yaml)

 [valid_config_blank.yaml](./tests/valid_config_blank.yaml)
 


CHANGELOG.md
# 0.4.22
- Added fallback support for special contacts in Slack (such as `_deployer`).
# 0.4.20
- Fixed crash when deleting subscriptions.
# 0.4.19
- Minor bug fixed: comparison of triggers with saturations that have parameters now works correctly.
# 0.4.18
- Saturation support was added.
# 0.4.17
- Added retrai requests to Moira.
# 0.4.16
- Bug fix with failed subscriptions
# 0.4.15
- Now you can configure inheritance for all triggers with the MONAD tag at once.
# 0.4.14
- Supported deletion of all triggers / subscriptions of one file
# 0.4.13
- Subscriptions already saved in other tags are no longer deleted.
# 0.4.12
- Created subscriptions are now also saved in Redis, like triggers.
- Added a link to the assembly in Teamcity in the user agent, in which alert-autoconf is running.
- Improved logs.
# 0.4.11
- {cluster} supported in targets and parents.tags fields
# 0.4.10
- Fixed validate.py
# 0.4.9
- Triggers now roll in two passes to avoid inheritance issues.
# 0.4.7
- Trigger inheritance supported
- Fixed a bug with updating targets, in which duplicate triggers were created instead of updating them.
# 0.4.6
- Schedules in subscriptions
# 0.4.5
- Warn_value and error_value fields must be filled together
# 0.4.4
- User Agent now contains the name and version of the utility
- Corrected documentation
# 0.4.3
- We exclude system tags when selecting triggers for comparison if there is no service token in Redis
# 0.4.2
- Fixed a bug in the comparison of triggers
# 0.4.1
- warning_value renamed to warn_value according to the trigger model in Moira
# 0.4.0
- In docker images, python version has been updated to 3.7
- Used for data validation pydantic
- alerting section made optional
# 0.3.12
- Processing changes in the fields day_disable, time_start, time_end
# 0.3.11
- Updated moira-client version to 2.4.2
# 0.3.10
- Contacts containing "{" are added to exceptions
# 0.3.9
- Fixed bug in validator: section `triggers` is now considered optional
# 0.3.8
- Added MONAD tag to exceptions
# 0.3.7
- Added alert.yaml validation script
# 0.3.6
- Added timeouts for moir client
# 0.3.5
- Fixed problem with prefixes
# 0.3.4
- Fixed problem with prefixes
# 0.3.3
- Fixed processing of dashboards links
# 0.3.2
- Changed the logic of processing triggers, in which changes are not recorded, we no longer update them, making a full scan in radishes!
# 0.3.1
- correct removal of triggers from Radish, when removing them in Moira by hand.
# 0.3.0
- Registration of incoming alert.yaml in Redis (argument "-s" with value "redis: // $ host: $ port / $ db_num").
  Exapmle: "-s redis: // monitoring01: 6379/5"
- Reconciliation of triggers from aletr.yaml with those in Redis (sset) (argument "-t" with value "$ teamCity_project_name").
  Exapmle: "-t service-saga"
- If alert.yaml is absent in the list of registered alert.yaml`s in Redis,
  triggers will be requested according to their tags.
# 0.2.8
- When managing subscriptions, added to exceptions, processing of system tags ERROR, OK, ...
# 0.2.7
- bugfix
# 0.2.6
- removed the addition of prefixes for system tags ERROR, WARN, OK, NODATA
# 0.2.5
- added parameter pending_interval
# 0.2.4
- added check for remote triggers and dashboards
# 0.2.3
- added a check for the presence of contacts when creating a subscription
# 0.2.2
- added escalations in subscriptions
- updated description of alert expressions
# 0.2.1
- Fixed url in entrypoint at http://moira.yourdomain.ru/api/
# 0.2.0
- New version of the config (version: "1.1") with support for `prefix` - an analogue of namespace for trigger names and tags.
# 0.1.5
- Fixed a bug update triggers (Work has been done to optimize the search and identify changed / deleted
  triggers)
# 0.1.4
- The algorithm for updating the subscription to the event has been changed
  When updating the list - those subscriptions that were installed in Moira and which do not change the configuration file in any way,
  I do not delete. Comparison is based on the list of contacts and subscription tags.
- The algorithm for updating triggers has been changed
  As well as in the subscription renewal algorithm - two lists of triggers are formed for comparison. First list
  generated from Moira's data, where triggers from the configuration file are mentioned. Second list for the most
  configuration file. Nothing is done on triggers that fall into the intersection of lists. With the rest
  triggers:
    - if they are from Moira's list, they are removed
    - if they are from the file list, they are added
- Added Dockerfile

# 0.1.3
- The triggers section in the configuration file is optional

# 0.1.2
- Fixed a bug with setting the logging level
- Fixed bug with installing jsonschema from package

# 0.1.1
- Fixed the algorithm for updating the subscription.
  Now, when updating the list of tagged subscriptions, the previous data will be completely replaced by the new ones.

# 0.1
- Implemented basic functionality


tests/test_alert.py
import logging

import uuid

from unittest import TestCase
from unittest.mock import patch
from unittest.mock import Mock

from pydantic import ValidationError

from moira_client import Moira
from moira_client.client import Client
from moira_client.models.tag import TagManager
from moira_client.models.trigger import TriggerManager, Trigger
from moira_client.models.contact import ContactManager, Contact
from moira_client.models.subscription import SubscriptionManager

from redis import Redis

from alert_autoconf.moira import MoiraAlert
from alert_autoconf.config import read_from_file
from alert_autoconf import LOG_FORMAT

logging.basicConfig(level=logging.ERROR, format=LOG_FORMAT)

logger = logging.getLogger('alert')


class MoiraAlertTest(TestCase):
    def assertDictKeysEqual(self, custom, expected):
        """
        Проверка ключей словарей
        :param expected: - словарь, который ожидаем
        :param custom: - словарь, который вернул запрос
        :return:
        """
        self.assertIsInstance(expected, dict, msg='expected are not dict')
        self.assertIsInstance(custom, dict, msg='custom are not dict')
        if expected.keys() == custom.keys():
            for k in expected:
                if isinstance(expected[k], dict):
                    self.assertDictKeysEqual(custom[k], expected[k])
        else:
            custom_error = expected.keys() ^ (expected.keys() | custom.keys())
            expected_error = custom.keys() ^ (expected.keys() | custom.keys())

            self.assertEqual(
                custom,
                set(),
                msg='Key(s) [{}] are not found in expected dictionary'.format(
                    custom_error
                ),
            )
            self.assertEqual(
                expected,
                set(),
                msg='Key(s) [{}] are not found in custom dictionary'.format(
                    expected_error
                ),
            )

    def setUp(self):
        self.url = 'http://localhost:1234/api/'
        self.token = 'test'
        self.redis_token_storage = 'redis://localhost:5678/10'

        self.moira = Moira(self.url)
        self.redis = Redis.from_url(self.redis_token_storage)

    def createCustomTriggers(self):
        current_triggers = {}
        trigger_id = str(uuid.uuid4())
        current_triggers[trigger_id] = Trigger(
            client=None,
            id=trigger_id,
            name='Trigger_1',
            targets=['stats.timer_1', 'stats.timer_2'],
            desc='Test description 1',
            warn_value=15,
            error_value=10,
            ttl=100,
            tags=['service_1'],
            ttl_state='ERROR',
            sched={
                'days': [{'enabled': False, 'name': 'Tue'}],
                'startOffset': 750,
                'endOffset': 1439,
            },
        )
        trigger_id = str(uuid.uuid4())
        current_triggers[trigger_id] = Trigger(
            client=None,
            id=trigger_id,
            name='Trigger_2',
            targets=['stats.timer_3', 'stats.timer4'],
            desc='Test description 2',
            warn_value=10,
            error_value=20,
            ttl=20,
            tags=['service_2'],
            ttl_state='NODATA',
            expression='(t1 > 10) ? ERROR : ((t2 > 4) ? WARN : OK)',
            sched={
                'days': [
                    {'enabled': False, 'name': 'Tue'},
                    {'enabled': False, 'name': 'Wed'},
                ],
                'startOffset': 750,
                'endOffset': 1439,
            },
        )

        return current_triggers

    def test_invalid_config_read_raises_ValidationError(self):
        invalid_config = 'tests/invalid_config.yaml'

        with self.assertRaises(ValidationError):
            read_from_file(invalid_config)

    def test_invalid_constructor_call_raises_TypeError(self):
        with self.assertRaises(TypeError):
            MoiraAlert('')

    def test_call_setup_with_incorrect_argument_raise_AttributeError(self):
        alert = MoiraAlert(self.moira, self.redis, self.token)

        with self.assertRaises(AttributeError):
            alert.setup('')

    def test_config_prefix(self):
        data = read_from_file('tests/valid_config_second.yaml')
        prefix = data.prefix
        for trigger in data.triggers:
            self.assertTrue(trigger.name.startswith(prefix), 'Trigger name with prefix')
            for tag in trigger.tags:
                self.assertTrue(tag.startswith(prefix), 'Trigger tag with prefix')

        for alerting in data.alerting:
            for tag in alerting.tags:
                self.assertTrue(tag.startswith(prefix), 'Alerting tag with prefix')

    def test_create_fake_moira_trigger_and_alerting_prefix_not_exists_token_not_exists(
        self
    ):
        valid_config = 'tests/valid_config.yaml'
        data = read_from_file(valid_config)
        client = Client(self.url)

        with patch.object(client, 'get', return_value={'list': []}), patch.object(
            client, 'put', return_value={}
        ):
            trigger_manager = TriggerManager(client)
            tag_manager = TagManager(client)
            contact_manager = ContactManager(client)
            subscription_manager = SubscriptionManager(client)

            with patch.object(
                self.moira, '_trigger', return_value=trigger_manager
            ) as _trigger_mock, patch.object(
                self.moira, '_contact', return_value=contact_manager
            ) as _contact_mock, patch.object(
                self.moira, '_tag', return_value=tag_manager
            ) as _tag_mock, patch.object(
                self.moira, '_subscription', return_value=subscription_manager
            ), patch.object(
                self.redis, 'exists', return_value=0
            ), patch.object(
                self.redis, 'sadd', return_value=1
            ) as _redis_sadd_mock:
                _contact_mock.fetch_by_current_user.return_value = []
                _contact_mock.add = Mock(
                    side_effect=lambda value, contact_type: Contact(
                        id=str(uuid.uuid4()), type=contact_type, value=value
                    )
                )
                trigger_orig = MoiraAlert(self.moira, self.redis, self.token)
                trigger_orig.setup(data)

            args, kwargs = _trigger_mock.create.call_args_list[0]
            kwargs.pop('name')
            count_contacts_in_alerting_block = self._get_count_key_by_dict(
                data.alerting, 'contacts'
            )
            count_contacts_in_alerting_block += 2  # + 2 contacts from escalations

            self.assertFalse(_tag_mock.fetch_assigned_triggers_by_tags.called)
            self.assertDictKeysEqual(
                {
                    k: v
                    for k, v in data.triggers[0].to_custom_dict().items()
                    if k not in ('name',)
                },
                kwargs,
            )
            self.assertTrue(_contact_mock.add.called)
            self.assertTrue(_trigger_mock.create.called)
            self.assertEqual(
                _contact_mock.add.call_count, count_contacts_in_alerting_block
            )
            self.assertEqual(_trigger_mock.create.call_count, len(data.triggers))
            self.assertEqual(_redis_sadd_mock.call_count, len(data.triggers))

    def test_create_fake_moira_trigger_and_alerting_prefix_exists_token_not_exists(
        self
    ):
        valid_config = 'tests/valid_config_second.yaml'
        data = read_from_file(valid_config)
        trigger_count = len(data.triggers)
        client = Client(self.url)

        with patch.object(client, 'get', return_value={'list': []}), patch.object(
            client, 'put', return_value={}
        ):
            trigger_manager = TriggerManager(client)
            tag_manager = TagManager(client)
            contact_manager = ContactManager(client)
            subscription_manager = SubscriptionManager(client)

            with patch.object(
                self.moira, '_trigger', return_value=trigger_manager
            ) as _trigger_mock, patch.object(
                self.moira, '_contact', return_value=contact_manager
            ) as _contact_mock, patch.object(
                self.moira, '_tag', return_value=tag_manager
            ) as _tag_mock, patch.object(
                self.moira, '_subscription', return_value=subscription_manager
            ), patch.object(
                self.redis, 'exists', return_value=0
            ) as _redis_exists, patch.object(
                self.redis, 'sadd', return_value=1
            ) as _redis_sadd:
                _contact_mock.fetch_by_current_user.return_value = []
                _contact_mock.add = Mock(
                    side_effect=lambda value, contact_type: Contact(
                        id=str(uuid.uuid4()), type=contact_type, value=value
                    )
                )
                trigger_orig = MoiraAlert(self.moira, self.redis, self.token)
                trigger_orig.setup(data)

            args, kwargs = _trigger_mock.create.call_args_list[0]
            kwargs.pop('name')
            count_contacts_in_alerting_block = self._get_count_key_by_dict(
                data.alerting, 'contacts'
            )
            count_contacts_in_alerting_block += 2  # + 2 contacts from escalations

            self.assertTrue(_tag_mock.fetch_assigned_triggers_by_tags.called)
            self.assertDictKeysEqual(
                {
                    k: v
                    for k, v in data.triggers[0].to_custom_dict().items()
                    if k not in ('name',)
                },
                kwargs,
            )
            self.assertTrue(_contact_mock.add.called)
            self.assertTrue(_trigger_mock.create.called)
            self.assertEqual(
                _contact_mock.add.call_count, count_contacts_in_alerting_block
            )
            self.assertEqual(_trigger_mock.create.call_count, trigger_count)
            self.assertEqual(_redis_exists.call_count, 1)
            self.assertEqual(_redis_sadd.call_count, trigger_count)

    def test_create_fake_moira_trigger_and_alerting_token_exists_and_empty(self):
        valid_config = 'tests/valid_config.yaml'
        data = read_from_file(valid_config)
        trigger_count = len(data.triggers)
        client = Client(self.url)

        with patch.object(client, 'get', return_value={'list': []}), patch.object(
            client, 'put', return_value={}
        ):
            trigger_manager = TriggerManager(client)
            tag_manager = TagManager(client)
            contact_manager = ContactManager(client)
            subscription_manager = SubscriptionManager(client)

            with patch.object(
                self.moira, '_trigger', return_value=trigger_manager
            ) as _trigger_mock, patch.object(
                self.moira, '_contact', return_value=contact_manager
            ) as _contact_mock, patch.object(
                self.moira, '_tag', return_value=tag_manager
            ) as _tag_mock, patch.object(
                self.moira, '_subscription', return_value=subscription_manager
            ), patch.object(
                self.redis, 'exists', return_value=1
            ) as _redis_exists_mock, patch.object(
                self.redis, 'smembers', return_value=set()
            ), patch.object(
                self.redis, 'sadd', return_value=1
            ) as _redis_sadd_mock:
                _contact_mock.fetch_by_current_user.return_value = []
                _contact_mock.add = Mock(
                    side_effect=lambda value, contact_type: Contact(
                        id=str(uuid.uuid4()), type=contact_type, value=value
                    )
                )
                trigger_orig = MoiraAlert(self.moira, self.redis, self.token)
                trigger_orig.setup(data)

            args, kwargs = _trigger_mock.create.call_args_list[0]
            kwargs.pop('name')
            count_contacts_in_alerting_block = self._get_count_key_by_dict(
                data.alerting, 'contacts'
            )
            count_contacts_in_alerting_block += 2  # + 2 contacts from escalations

            self.assertFalse(_tag_mock.fetch_assigned_triggers_by_tags.called)
            self.assertDictKeysEqual(
                {
                    k: v
                    for k, v in data.triggers[0].to_custom_dict().items()
                    if k not in ('name',)
                },
                kwargs,
            )
            self.assertTrue(_contact_mock.add.called)
            self.assertFalse(_trigger_mock.fetch_by_id.called)
            self.assertTrue(_trigger_mock.create.called)
            self.assertEqual(
                _contact_mock.add.call_count, count_contacts_in_alerting_block
            )
            self.assertEqual(_trigger_mock.create.call_count, trigger_count)
            self.assertTrue(_redis_exists_mock.called)
            self.assertEqual(_redis_sadd_mock.call_count, trigger_count)

    def test_create_fake_moira_trigger_and_alerting_token_exists_and_not_empty(self):
        valid_config = 'tests/valid_config.yaml'
        data = read_from_file(valid_config)
        current_triggers = self.createCustomTriggers()
        moira_triggers = dict([current_triggers.popitem()])
        client = Client(self.url)

        redis_trigger_ids = set(i.encode('utf-8') for i in moira_triggers.keys())
        redis_trigger_ids = redis_trigger_ids | {str(uuid.uuid4()).encode('utf-8')}

        with patch.object(client, 'get', return_value={'list': []}), patch.object(
            client, 'put', return_value={}
        ):
            trigger_manager = TriggerManager(client)
            tag_manager = TagManager(client)
            contact_manager = ContactManager(client)
            subscription_manager = SubscriptionManager(client)

            with patch.object(
                self.moira, '_trigger', return_value=trigger_manager
            ) as _trigger_mock, patch.object(
                self.moira, '_contact', return_value=contact_manager
            ) as _contact_mock, patch.object(
                self.moira, '_tag', return_value=tag_manager
            ) as _tag_mock, patch.object(
                self.moira, '_subscription', return_value=subscription_manager
            ), patch.object(
                self.redis, 'exists', return_value=1
            ) as _redis_exists_mock, patch.object(
                self.redis, 'smembers', return_value=redis_trigger_ids
            ) as _redis_smembers_mock, patch.object(
                self.redis, 'sadd', return_value=1
            ) as _redis_sadd_mock, patch.object(
                self.redis, 'srem', return_value=1
            ) as _redis_srem_mock:
                _trigger_mock.fetch_by_id = Mock(
                    side_effect=lambda id: moira_triggers.get(id, None)
                )
                _contact_mock.fetch_by_current_user.return_value = []
                _contact_mock.add = Mock(
                    side_effect=lambda value, contact_type: Contact(
                        id=str(uuid.uuid4()), type=contact_type, value=value
                    )
                )
                trigger_orig = MoiraAlert(self.moira, self.redis, self.token)
                trigger_orig.setup(data)

            args, kwargs = _trigger_mock.create.call_args_list[0]
            kwargs.pop('name')
            count_contacts_in_alerting_block = self._get_count_key_by_dict(
                data.alerting, 'contacts'
            )
            count_contacts_in_alerting_block += 2  # + 2 contacts from escalations

            self.assertFalse(_tag_mock.fetch_assigned_triggers_by_tags.called)
            self.assertDictKeysEqual(
                {
                    k: v
                    for k, v in data.triggers[0].to_custom_dict().items()
                    if k not in ('name',)
                },
                kwargs,
            )
            self.assertTrue(_contact_mock.add.called)
            self.assertEqual(_trigger_mock.fetch_by_id.call_count, 2)
            self.assertTrue(_trigger_mock.create.called)
            self.assertEqual(
                _contact_mock.add.call_count, count_contacts_in_alerting_block
            )
            self.assertEqual(_trigger_mock.create.call_count, 1)
            self.assertTrue(_redis_exists_mock.call_count)
            self.assertTrue(_redis_smembers_mock.call_count)
            self.assertEqual(_redis_sadd_mock.call_count, 1)
            self.assertEqual(_redis_srem_mock.call_count, 1)

    def _get_count_key_by_dict(self, data: list, key: str) -> int:
        cnt = 0
        for alert in data:
            contacts = alert.__getattribute__(key)
            if contacts:
                cnt += len(contacts)
        return cnt


alert_autoconf/models.py
from uuid import UUID
from datetime import time
from enum import Enum
from typing import Dict, List, Optional
from pydantic import BaseModel, AnyHttpUrl, validator, root_validator


class ContactTypeEnum(Enum):
    JIRA = "jira"
    MAIL = "mail"
    PUSHOVER = "pushover"
    SEND_SMS = "send-sms"
    SLACK = "slack"
    TELEGRAM = "telegram"
    TWILIO_SMS = "twilio sms"
    TWILIO_VOICE = "twilio voice"


class DaysEnum(Enum):
    MON = "Mon"
    TUE = "Tue"
    WED = "Wed"
    THU = "Thu"
    FRI = "Fri"
    SAT = "Sat"
    SUN = "Sun"


class TtlStateEnum(Enum):
    DEL = "DEL"
    ERROR = "ERROR"
    NODATA = "NODATA"
    OK = "OK"
    WARN = "WARN"


class ParentTriggerRef(BaseModel):
    tags: List[str]
    name: str

    def __hash__(self):
        return hash((
            frozenset(tags),
            name,
        ))

    def __eq__(self, other):
        return (
            set(self.tags) == set(other.tags)
            and self.name == other.name
        )


class Saturation(BaseModel):
    type: str
    fallback: Optional[str] = None
    parameters: Optional[dict] = None

    def to_custom_dict(self) -> Dict:
        result = {
            "type": self.type,
        }
        if self.fallback is not None:
            result["fallback"] = self.fallback
        if self.parameters is not None:
            result["extra_parameters"] = self.parameters
        return result

    @classmethod
    def from_moira_client_model(cls, moira_saturation: "moira_client.models.trigger.Saturation"):
        d = moira_saturation.to_dict()
        d["parameters"] = d.pop("extra_parameters", None)
        return cls(**d)

    def __hash__(self):
        dct = self.to_custom_dict()
        return hash(_freeze_dict(dct))

    def __eq__(self, other):
        if isinstance(other, Saturation):
            return self.to_custom_dict() == other.to_custom_dict()
        else:
            raise ValueError("Incomparable types")


def _freeze_dict(dct):
    """Tries to freeze a dict to make it hashable."""
    result = []
    for key, value in dct.items():
        if isinstance(value, dict):
            value = _freeze_dict(value)
        result.append((key, value))
    result.sort()
    return tuple(result)


class Trigger(BaseModel):
    id: Optional[str] = None
    name: str
    tags: List[str]
    targets: List[str]
    warn_value: Optional[int] = None
    error_value: Optional[int] = None
    desc: str = ""
    ttl: int = 600
    ttl_state: TtlStateEnum = TtlStateEnum.NODATA
    expression: Optional[str] = ""

    is_pull_type: bool = False
    dashboard: Optional[AnyHttpUrl] = None
    pending_interval: Optional[int] = 0

    day_disable: List[DaysEnum] = []
    time_start: Optional[time] = time(hour=0, minute=0)
    time_end: Optional[time] = time(hour=23, minute=59)

    parents: Optional[List[str]]

    saturation: Optional[List[Saturation]] = list()

    @validator("id")
    def id_uuid(cls, v):
        try:
            UUID(v)
        except ValueError:
            raise
        return v

    @root_validator
    def check_thresholds_values(cls, values):
        warn_value, error_value = (
            values.get('warn_value') is not None,
            values.get('error_value') is not None,
        )
        if warn_value ^ error_value:
            raise ValueError('must provide warn_value and error_value')

        if (
            warn_value & error_value
            and len(values.get('targets')) > 1
            and values.get('expression') is None
        ):
            raise ValueError('must use single target with warn_value and error_value')

        return values

    def to_custom_dict(self) -> Dict:
        return {
            'name': self.name,
            'tags': self.tags,
            'targets': self.targets,
            'warn_value': self.warn_value,
            'error_value': self.error_value,
            'desc': self.desc,
            'ttl': self.ttl,
            'ttl_state': self.ttl_state.value,
            'expression': self.expression,
            'is_pull_type': self.is_pull_type,
            'dashboard': self.dashboard,
            'pending_interval': self.pending_interval,
            'sched': {
                'startOffset': self.time_start.hour * 60 + self.time_start.minute,
                'endOffset': self.time_end.hour * 60 + self.time_end.minute,
                'tzOffset': 0,
                'days': [
                    {'name': day.value, 'enabled': day not in self.day_disable}
                    for day in DaysEnum
                ],
            },
            'parents': self.parents,
            'saturation': [
                s.to_custom_dict()
                for s in self.saturation
            ],
        }


class TriggerFile(Trigger):
    parents: Optional[List[ParentTriggerRef]]


class Contact(BaseModel):
    id: Optional[str] = None
    type: ContactTypeEnum
    value: str
    fallback_value: Optional[str] = None

    def __hash__(self):
        return f"{self.type}:{self.value}:{self.fallback_value}".__hash__()


class Escalation(BaseModel):
    contacts: List[Contact]
    offset_in_minutes: int = 0


class Subscription(BaseModel):
    tags: List[str]
    contacts: Optional[List[Contact]] = []
    escalations: Optional[List[Escalation]] = []
    day_disable: List[DaysEnum] = []
    time_start: Optional[time] = time(hour=0, minute=0)
    time_end: Optional[time] = time(hour=23, minute=59)

    def to_custom_dict(self) -> Dict:
        return {
            'tags': self.tags,
            'contacts': [c.id for c in self.contacts],
            'escalations': [
                {
                    'contacts': [c.id for c in e.contacts],
                    'offset_in_minutes': e.offset_in_minutes,
                }
                for e in self.escalations
            ],
            'sched': {
                'startOffset': self.time_start.hour * 60 + self.time_start.minute,
                'endOffset': self.time_end.hour * 60 + self.time_end.minute,
                'tzOffset': 0,
                'days': [
                    {'name': day.value, 'enabled': day not in self.day_disable}
                    for day in DaysEnum
                ],
            },
        }


class Alerts(BaseModel):
    version: float = 1
    prefix: str = ""
    triggers: List[TriggerFile] = []
    alerting: List[Subscription] = []
